### PR TITLE
kron for Diagonal uses kron for components

### LINKED ISSUE #40595: kron for Diagonal matrices explicitly converts to Vector losing static size information
Ideally the Kronecker product of `StaticArray`s would remain static. However

```julia
julia> kron(Diagonal(SVector{3}(1,2,3)), Diagonal(SVector{3}(1,2,3)))
9×9 Diagonal{Int64, Vector{Int64}}:
 1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  2  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  3  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  2  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  4  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  6  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  3  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  6  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  9
```

This happens because https://github.com/JuliaLang/julia/blob/1474566ffc1dfe16f643b446f667178b78b56aa6/stdlib/LinearAlgebra/src/diagonal.jl#L496-L502 constructs a `Vector` explicitly.

Perhaps this `kron` method could internally call `kron` on the diagonal vectors? This appears to be defined to preserve the sizes:
```julia
julia> kron(SVector{3}(1,2,3), SVector{3}(1,2,3))
9-element SVector{9, Int64} with indices SOneTo(9):
 1
 2
 3
 2
 4
 6
 3
 6
 9
```

### PR DESCRIPTION
**Pull request by jishnub**
Fixes #40595, but perhaps needs more tests as the specific case mentioned in the issue is not tested yet. I'm not sure how to add tests for static sizes, suggestions welcome.

After this PR:
```julia
julia> kron(Diagonal(SVector{3}(1,2,3)), Diagonal(SVector{3}(1,2,3)))
9×9 Diagonal{Int64, SVector{9, Int64}}:
 1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  2  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  3  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  2  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  4  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  6  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  3  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  6  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  9
```

### CODE REVIEW

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -494,11 +494,9 @@ rdiv!(A::AbstractMatrix{T}, transD::Transpose{<:Any,<:Diagonal{T}}) where {T} =
...
 function kron(A::Diagonal{T1}, B::Diagonal{T2}) where {T1<:Number, T2<:Number}
-    valA = A.diag; nA = length(valA)
-    valB = B.diag; nB = length(valB)
-    valC = Vector{typeof(zero(T1)*zero(T2))}(undef,nA*nB)
-    C = Diagonal(valC)
-    return @inbounds kron!(C, A, B)
+    Cdiag = kron(A.diag, B.diag)
+    C = Diagonal(Cdiag)
+    return C
 end
```
Comment (johnnychen94) [inline stdlib/LinearAlgebra/src/diagonal.jl:500]: Could be simplified to

```suggestion
kron(A::Diagonal{<:Number}, B::Diagonal{<:Number}) = Diagonal(kron(A.diag, B.diag))
```

**Review by johnnychen94 — COMMENTED** [review]
> I'm not sure how to add tests for static sizes, suggestions welcome.

As long as the diag array type not changed, I think we're good here. So maybe test it with a custom array type:

```julia
struct KronTestArray{T, N, AT} <: AbstractArray{T, N}
    data::AT
end
KronTestArray(data::AbstractArray) = KronTestArray{eltype(data), ndims(data), typeof(data)}(data)
Base.size(A::KronTestArray) = size(A.data)

Base.kron(A::KronTestArray, B::KronTestArray) = KronTestArray(kron(A.data, B.data))
```

```julia
julia> A = KronTestArray([1, 2, 3]);

julia> kron(A, A) isa KronTestArray
true

julia> Ad = Diagonal(A);

julia> kron(Ad, Ad).diag isa KronTestArray
true
```

File: stdlib/LinearAlgebra/src/diagonal.jl (hunk 1 above)
Comment (jishnub) [reply stdlib/LinearAlgebra/src/diagonal.jl]: Updated to use this form

File: stdlib/LinearAlgebra/test/diagonal.jl [hunk 2]
```diff
@@ -403,6 +403,21 @@ Random.seed!(1)
...
+        data::AT
+    end
+    KronTestArray(data::AbstractArray) = KronTestArray{eltype(data), ndims(data), typeof(data)}(data)
+    Base.size(A::KronTestArray) = size(A.data)
+    LinearAlgebra.kron(A::KronTestArray, B::KronTestArray) = KronTestArray(kron(A.data, B.data))
+
+    A = KronTestArray([1, 2, 3]);
+    @test kron(A, A) isa KronTestArray
+    Ad = Diagonal(A);
+    @test kron(Ad, Ad).diag isa KronTestArray
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/test/diagonal.jl:418]: Just to make sure we get the correct values:
```suggestion
    @test kron(Ad, Ad).diag isa KronTestArray
    @test kron(Ad, Ad).diag == kron([1, 2, 3], [1, 2, 3])
```
Not sure if we have that covered elsewhere.

**Review by dkarrasch — APPROVED** [review]
A little addition, otherwise LGTM.

**Pull request merged**